### PR TITLE
Add fork branch sync

### PR DESCRIPTION
**Pull request by jonesbusy**
# Description

Add support for fork sync branch

Fixing https://github.com/hub4j/github-api/issues/1626

Added two wiremock test and tested a SNAPSHOT version in a WIP PR https://github.com/jenkinsci/plugin-modernizer-tool/pull/141

I confirm this work by enabling logs and ensuring my fork is sync directly on GitHub

```
(e24ddd0) GitHub API request: POST https://api.github.com/repos/jonesbusy/jobcacher-plugin/merge-upstream 
```

# Before submitting a PR:

- [x] Changes must not break binary backwards compatibility. If you are unclear on how to make the change you think is needed while maintaining backward compatibility, [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [x] Add JavaDocs and other comments explaining the behavior. 
- [x] When adding or updating methods that fetch entities, add `@link` JavaDoc entries to the relevant documentation on https://docs.github.com/en/rest . 
- [x] Add tests that cover any added or changed code. This generally requires capturing snapshot test data. See [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [x] Run `mvn -D enable-ci clean install site` locally. If this command doesn't succeed, your change will not pass CI.
- [x] Push your changes to a branch other than `main`. You will create your PR from that branch.

# When creating a PR: 

  - [x] Fill in the "Description" above with clear summary of the changes. This includes:
  - [x] If this PR fixes one or more issues, include "Fixes #<issue number>" lines for each issue. 
  - [x] Provide links to relevant documentation on https://docs.github.com/en/rest where possible. If not including links, explain why not.
- [x] All lines of new code should be covered by tests as reported by code coverage. Any lines that are not covered must have PR comments explaining why they cannot be covered. For example, "Reaching this particular exception is hard and is not a particular common scenario."
- [x] Enable "Allow edits from maintainers".

API is https://docs.github.com/en/rest/branches/branches?apiVersion=2022-11-28#sync-a-fork-branch-with-the-upstream-repository

Code is covered 100%

Screenshot from 2024-07-24 12-38-45

### CODE REVIEW

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 1]
```diff
@@ -1614,6 +1614,27 @@ public GHRepository fork() throws IOException {
...
+     * Sync this repository fork branch
+     *
+     * @param branch
+     *            the branch to sync
+     * @return The current repository
+     * @throws IOException
+     *             the io exception
+     */
+    public GHRepository sync(String branch) throws IOException {
+        if (isFork()) {
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHRepository.java:1627]: If someone tries to sync a repo that is not a fork, shouldn't we throw an exception?  Would the REST API throw an exception if this call were made against a repo that is not a fork?

Comment (jonesbusy) [reply src/main/java/org/kohsuke/github/GHRepository.java:1627]: Hi,

Yes I get 422

```
org.kohsuke.github.HttpException: {"message":"Validation Failed","documentation_url":"https://docs.github.com/rest/branches/branches#sync-a-fork-branch-with-the-upstream-repository","status":"422"}
```

What do you prefer ? Remove the check or let the exception to the user ?

Comment (jonesbusy) [reply src/main/java/org/kohsuke/github/GHRepository.java]: Done in 5083df701414897ed171202aba5ea262a589d5e3

File: src/test/java/org/kohsuke/github/GHRepositoryTest.java [hunk 2]
```diff
@@ -47,6 +47,18 @@ private GHRepository getRepository(GitHub gitHub) throws IOException {
...
+    /**
+     * Test sync of fork
+     *
+     * @throws IOException
+     *             Signals that an I/O exception has occurred.
+     */
+    @Test
+    public void sync() throws IOException {
+        GHRepository r = getRepository();
+        r.sync("main");
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GHRepositoryTest.java:59]: Add slightly more complex tests to cover the returned record and also the error/exception case.  (You can do all in one method if you want or separately.)

**Review by bitwiseman — CHANGES_REQUESTED** [review]
Second rounds of feedback, sorry.  This is looking good, just a little more to be done.  
Thanks!

**Comment by jonesbusy** [comment]
I've added the `GHBranchSync` model and return it

I've also added a negative test when the API return a non-200 status code (for example 422 when the repo is not a fork)

**Review by gounthar — APPROVED** [review]
Nice, thanks! 👍

**Review by NotMyFault — APPROVED** [review]
Awesome work!

File: src/test/java/org/kohsuke/github/GHRepositoryTest.java [hunk 3]
```diff
@@ -47,6 +47,36 @@ private GHRepository getRepository(GitHub gitHub) throws IOException {
...
+     * Test sync of repository not a fork
+     *
+     * @throws IOException
+     *             Signals that an I/O exception has occurred.
+     */
+    @Test(expected = HttpException.class)
+    public void syncNoFork() throws IOException {
+        GHRepository r = getRepository();
+        GHBranchSync sync = r.sync("main");
+        fail("Should have thrown an exception");
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GHRepositoryTest.java:76]: There's an `assertThrows()` method that I prefer for this, but I'm going to block merge for it.

File: src/main/java/org/kohsuke/github/GHBranchSync.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    private GHRepository owner;
+
+    private String message;
+
+    @JsonProperty("merge_type")
+    private String mergeType;
+
+    @JsonProperty("base_branch")
+    private String baseBranch;
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHBranchSync.java:16]: The object mapper translates camel-case to snake-case automatically. 

```suggestion
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

public class GHBranchSync extends GitHubInteractiveObject {

    private GHRepository owner;

    private String message;

    private String mergeType;

    private String baseBranch;
```

**Review by bitwiseman — APPROVED** [review]
Great work on this. Thanks!

**Comment by jonesbusy** [comment]
Apparently something wrong with javadoc. I will take a look

**Comment by jonesbusy** [comment]
Should be fixed, can you restart the workflows please ?

**Comment by jonesbusy** [comment]
Hi @bitwiseman Sorry for the ping, but where you able to review again this PR ?

It's also update with the default branch

Thanks!

**Comment by jonesbusy** [comment]
Thanks for the merge @bitwiseman. Sorry to ping you directly (again) but is there a release soon including this feature ?

I'm asking because we are approaching the end of the Google Summer of Code on our project https://github.com/jenkinsci/plugin-modernizer-tool and would be great that we can include the new release on our delivery

The fork sync is quiet critical on our project to avoid opening PRs from outdated forks.

Thanks for your consideration 😃

**Comment by bitwiseman** [comment]
I'll do a release today.

**Pull request merged**